library/core/src/main/java/com/google/android/exoplayer2/source/ConcatenatingMediaSource.java

```diff
@@ -297,6 +304,79 @@ public final synchronized void removeMediaSource(
+   * @param toIndex The final range index, pointing to the first media source that will be left
+   *     untouched. The last media source to be removed is at index {@code toIndex} - 1.
+   *     This index must be in the range of 0 &lt;= index &lt; {@link #getSize()}.
+   *
+   * @throws IndexOutOfBoundsException when either index is out of playlist bounds, i.e. {@code
+   *      fromIndex} &lt; 0 or &gt;= {@link #getSize()}, {@code toIndex} &lt; 0 or &gt; {@link
+   *      #getSize()}
+   * @throws IndexOutOfBoundsException when range is malformed, i.e. {@code fromIndex} &gt;
+   *      {@code toIndex}
+   */
+  public final synchronized void removeMediaSourceRange(int fromIndex, int toIndex) {
+    removeMediaSourceRange(fromIndex, toIndex, null);
+  }
+
+  /**
+   * Removes a range of {@link MediaSource}s from the playlist, by specifying an initial index
+   * (included) and a final index (excluded), and executes a custom action on completion..
+   *
+   * <p>Note: when specified range is empty, no actual media source is removed and no exception
+   * is thrown.
+   *
+   * @param fromIndex The initial range index, pointing to the first media source that will be
+   *     removed. This index must be in the range of 0 &lt;= index &lt; {@link #getSize()}.
+   * @param toIndex The final range index, pointing to the first media source that will be left
+   *     untouched. The last media source to be removed is at index {@code toIndex} - 1.
+   *     This index must be in the range of 0 &lt;= index &lt; {@link #getSize()}.
+   *
+   * @throws IndexOutOfBoundsException when either index is out of playlist bounds, i.e. {@code
+   *      fromIndex} &lt; 0 or &gt;= {@link #getSize()}, {@code toIndex} &lt; 0 or &gt; {@link
+   *      #getSize()}
+   * @throws IndexOutOfBoundsException when range is malformed, i.e. {@code fromIndex} &gt;
+   *      {@code toIndex}
+   * @param actionOnCompletion A {@link Runnable} which is executed immediately after the media
+   *     source range has been removed from the playlist.
+   */
+  public final synchronized void removeMediaSourceRange(
+          int fromIndex, int toIndex, @Nullable Runnable actionOnCompletion) {
+    if (fromIndex < 0 || fromIndex >= mediaSourcesPublic.size()) {
+      throw new IndexOutOfBoundsException(String.format("Cannot remove source range: initial index (%d) out of bounds", fromIndex));
+    }
+    if (toIndex < 0 || toIndex > mediaSourcesPublic.size()) {
+      throw new IndexOutOfBoundsException(String.format("Cannot remove source range: final index (%d) out of bounds", toIndex));
+    }
+    if (fromIndex > toIndex) {
+      throw new IndexOutOfBoundsException(String.format("Cannot remove source range: range malformed (%d, %d)", fromIndex, toIndex));
+    }
+    if (fromIndex == toIndex) {
+      if (actionOnCompletion != null) {
+        actionOnCompletion.run();
+      }
+      return;
+    }
+    mediaSourcesPublic.subList(fromIndex, toIndex).clear();
```

Comment:
Can you replace that with Util.removeRange(mediaSourcesPublic, fromIndex, toIndex)? That calls the same method but is slightly more readable.

Also move it above the if(fromIndex == toIndex) check to ensure the exceptions are thrown first.

Reply:
Ok, just not sure what that Util method does w.r.t. exceptions (I think `sublist()` does not throw when indices are the same).